### PR TITLE
Question index

### PR DESCRIPTION
**Pull request by annaswigart**
Not ready to merge yet, but feel free to take a look, @salspaugh @macfarlandian .

Still working on getting the variable references. Just displays the id right now for each variable that is tied to the question. Any tips on the best way to query the variable (column) name?

Thoughts on whether we want to have a modal to add questions, or to allow the user to type the question in on the page, as we had in v1? A modal could include text, an indicator of keep/answerable, and maybe notes, but I think we could probably accomplish this without a modal. I'm leaning toward no modal.

Will also need styling/layout updates.

### CODE REVIEW

**Comment by macfarlandian** [comment]
1. i agree, no modal. 
2. to get variable info i would query the Columns collection with those IDs and limit the fields returned to "name" or whatever that field is called, and maybe the datatype also. to keep documents small i refactored all these relationships into references, and since the database is cached in memory anyway doing a bunch of queries doesn't add a lot of latency.
3. i noticed that when I add a new question it shows up for all the datasets, so the controller is either not attaching it to the dataset or not filtering the query properly

otherwise, coming along!

**Comment by annaswigart** [comment]
Hey @macfarlandian, @salspaugh 

I've made some progress on the questionIndex page. Can you take a look and let me know if this is on the right track?

Design issues that need feedback:
- thoughts on the icons?
- thoughts on wording? (I'm looking for another term for "Discard" in particular)
- I'm thinking of switching the edit dropdown to an edit mode toggle. In edit mode, maybe you could switch bins, change text, or delete the question.
- might get to drag and drop moving to change bins later (maybe not this week)

Code issues I need help with: 
- All questions across datasets show up when you first go into a dataset. If you hit refresh on your browser page, the questions that are not from the current dataset get filtered out. @macfarlandian, I tried for a while to fix this, but am stuck.
- The question and notes text is editable, but it does not save in the database. I'm thinking that I shouldn't use $scope, but something else. Thoughts?

**Comment by macfarlandian** [comment]
@annaswigart this looks awesome! 

**icons:** i don't really get what the "go to chart view" button is for ... is this what takes you to the view where you can add variables to a question? if so i think this action needs to feel more primary.

also the check/x/? can probably be with the bin labels instead of repeated in every card. makes the cards less cluttered. 

**wording**: brainstorming some alternatives: 
- keep/reject/undecided
- interesting/not interesting/not sure yet
- pursue/don't pursue/not sure yet
- explore/ignore/not sure (this one has the benefit of rhyming)

**edit mode**: personally i wouldn't bother, this works fine. the only thing that's not really very discoverable is that the question text is editable, but something like a hover effect could fix that. or even some kind of prompt in the edit menu (the pencil icon may make people think it lets them change the actual text also). 

**drag and drop**: nice to have but not essential IMO. alternatively what about breaking out the bin-change controls into another menu with an icon something like this for moving things around: http://fortawesome.github.io/Font-Awesome/icon/exchange/

**code stuff**: will address the dataset stuff separately (since you're stuck maybe i'll try to get it working and push to this PR?) - but for saving stuff to the database, i think the problem is that "$scope.question" is not a thing in the questionIndex view because there's no question id in the state params ... you need to pass it in from the ng-repeat loop somehow

**Comment by salspaugh** [comment]
I agree, this looks amazing! Following @macfarlandian's lead:

**Icons:** And also font -- this is a very small thing that we can easily change but FWIW I have been using grey or black for words and icons because we use "blue" to mean "string datatype" for the variable names and charts. Re: go to chart view -- my interpretation is this would take you to the single question view. Not sure what @macfarlandian means by "more primary". It feels pretty noticeable to me. But I think the mouseover text "Go to chart view" is a bit confusing.

**Wording:** All the suggestions made and what you have seem about the same quality to me. The only difference I would make is I would advocate considering multi-word options rather than single words to make it more clear, e.g., "Keep for later analysis" and "Discarded questions" or something. Exactly what the main verbs should be, I'm not sure.

**Edit mode:** Agree with @macfarlandian. Making it a little bit more clear you can edit by clicking would be nice, but I love the way the edit works now. The pencil icon is a bit confusing though especially because it says "Edit question or bin" in the mouseover, but it seems you can really only edit the bins. Using a different icon and mouseover text for that would help.

**Drag and drop:** I like the icon @macfarlandian suggested for moving bins! Better than the pencil. I don't think we should have both a "Go to charts view" icon and a "Go to charts view" item in the drop-down. I'd remove it from the drop-down. It would also be good to break "Delete" out into it's own button with a trashcan icon. Let's not worry about the drag and drop functionality until after we get more basics working.

**Code stuff:** I am not sure how to address your questions, but I helpfully seem to have found another bug: When you select "Go to chart view" from the drop down, it appears to delete the question.

**Comment by salspaugh** [comment]
If the "Go to chart view" button and menu item are meant to take you to the single question page, better wording might be "View question details".

**Comment by macfarlandian** [comment]
@annaswigart i think i got the "all questions" bug figured out ... please pull and confirm

**Comment by annaswigart** [comment]
Thanks so much for the feedback, @salspaugh and @macfarlandian!

I think I implemented all of the changes you guys suggested. I did keep the check mark/?/X icons on each question though because it brought attention to the status of each question and the bins just kind of looked like big text blobs (hard to parse) without them. Can reconsider though, if we find a better way to do this.

Suggestions for hover text that makes more sense than "go to chart view"? I contemplated making that functionality bigger/easier to discover somehow, but I think on this page we really want to encourage people to look at question text and create as many good questions as they can-- not be taken to the chart view automatically.

I also added a question alert to the garbage can (in an app-wide directive). It uses the default (ugly) browser alert box for now, but we can switch this out for something nicer. @macfarlandian do you have any thoughts on the best way to do this? I looked into a few options but am not sure. 

Please take a look and let me know if it's ready to merge!

**Comment by salspaugh** [comment]
This looks almost ready to merge except that the notes don't appear editable -- is that intentional? I feel like I want to be able to edit them there. We can always wait until after the usability tests to do this though. I think for now: :+1: The alert button thing was pretty unnoticeable to me not because it's subtly different (it is very different and pretty ugly) but just because I'm so used to seeing those that I hardly noticed.

**Comment by macfarlandian** [comment]
second sara's points. and would add that edits to question text still
aren't being saved. but I'll merge this and we can fix that later.

**Comment by annaswigart** [comment]
Right-- have not been able to resolve issue with question and notes text edits saving. The notes text is editable in my view, but it is hard to discover and the cursor is annoyingly in an odd place when you start typing. I'll merge in a bit-- in meetings for the next couple of hours.

**Comment by macfarlandian** [comment]
i can take care of the merge ... there are some conflicts with sara's code that i'm already resolving.

do you need this on the server for your usability tests? or will you just do that locally on one of your laptops?

**Comment by annaswigart** [comment]
Thanks! local is fine.
